{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "83bfad8b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import re\n",
    "import os\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from scipy import stats\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.linear_model import Lasso\n",
    "from sklearn.linear_model import Ridge\n",
    "from sklearn.model_selection import train_test_split\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2b6b0e43",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# load the dataset\n",
    "train_file=\"D:/PhD/My Papers/Paper2/interdisciplinary science_Revised/code/Data/processedData_GSE128235.csv\"\n",
    "label_file=\"D:/PhD/My Papers/Paper2/interdisciplinary science_Revised/code/Data/phenoData_GSE128235.csv\"\n",
    "platform=\"D:/PhD/My Papers/Paper2/interdisciplinary science_Revised/code/Data/probeGene_GSE128235.csv\"\n",
    "\n",
    "data = pd.read_csv(train_file, index_col=0)\n",
    "pheno = pd.read_csv(label_file, index_col=0)\n",
    "probeMap = pd.read_csv(platform, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dd053152",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "code=\"GSE128235\"\n",
    "#probeMap to dictionary\n",
    "probeMapDict={k: list(v) for k,v in probeMap.groupby(\"UCSC_RefGene_Name\")[\"Name\"]}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b05a881",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "data=data.dropna()\n",
    "data=data.dropna(axis=1)\n",
    "data=data.drop_duplicates(keep= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "162dd9b5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>geo_accession</th>\n",
       "      <th>status</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>last_update_date</th>\n",
       "      <th>type</th>\n",
       "      <th>channel_count</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>characteristics_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>cellcount.gran.ch1</th>\n",
       "      <th>cellcount.mono.ch1</th>\n",
       "      <th>cellcount.nk.ch1</th>\n",
       "      <th>diagnosis.ch1</th>\n",
       "      <th>education.level.ch1</th>\n",
       "      <th>history.of.myocardial.infarction.ch1</th>\n",
       "      <th>Sex.ch1</th>\n",
       "      <th>snp.based.principle.component.1.ch1</th>\n",
       "      <th>snp.based.principle.component.2.ch1</th>\n",
       "      <th>tissue.ch1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3667913</th>\n",
       "      <td>sample1</td>\n",
       "      <td>GSM3667913</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.57450</td>\n",
       "      <td>0.11560</td>\n",
       "      <td>0.01708</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>-0.022566</td>\n",
       "      <td>0.016230</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3667872</th>\n",
       "      <td>sample10</td>\n",
       "      <td>GSM3667872</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.50000</td>\n",
       "      <td>0.06131</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>-0.001741</td>\n",
       "      <td>-0.018000</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3667947</th>\n",
       "      <td>sample100</td>\n",
       "      <td>GSM3667947</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.52666</td>\n",
       "      <td>0.10827</td>\n",
       "      <td>0.00085</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>0.007919</td>\n",
       "      <td>0.003596</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3667878</th>\n",
       "      <td>sample101</td>\n",
       "      <td>GSM3667878</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.61128</td>\n",
       "      <td>0.16026</td>\n",
       "      <td>0.02816</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>0.014514</td>\n",
       "      <td>0.005688</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3667998</th>\n",
       "      <td>sample102</td>\n",
       "      <td>GSM3667998</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.53323</td>\n",
       "      <td>0.05734</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>-0.000620</td>\n",
       "      <td>-0.007328</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3667830</th>\n",
       "      <td>sample95</td>\n",
       "      <td>GSM3667830</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.67893</td>\n",
       "      <td>0.06394</td>\n",
       "      <td>0.02600</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>0.011203</td>\n",
       "      <td>0.002856</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3667866</th>\n",
       "      <td>sample96</td>\n",
       "      <td>GSM3667866</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.49627</td>\n",
       "      <td>0.06016</td>\n",
       "      <td>0.05033</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>0.014452</td>\n",
       "      <td>-0.001970</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3667815</th>\n",
       "      <td>sample97</td>\n",
       "      <td>GSM3667815</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.76167</td>\n",
       "      <td>0.04776</td>\n",
       "      <td>0.02331</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>-0.003758</td>\n",
       "      <td>0.019465</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3668001</th>\n",
       "      <td>sample98</td>\n",
       "      <td>GSM3668001</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.72225</td>\n",
       "      <td>0.08636</td>\n",
       "      <td>0.02902</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>-0.000920</td>\n",
       "      <td>-0.010731</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3667717</th>\n",
       "      <td>sample99</td>\n",
       "      <td>GSM3667717</td>\n",
       "      <td>Public on Mar 14 2019</td>\n",
       "      <td>Mar 13 2019</td>\n",
       "      <td>Mar 15 2019</td>\n",
       "      <td>genomic</td>\n",
       "      <td>1</td>\n",
       "      <td>whole blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>diagnosis: case</td>\n",
       "      <td>...</td>\n",
       "      <td>0.74141</td>\n",
       "      <td>0.06519</td>\n",
       "      <td>0.04442</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>-0.009914</td>\n",
       "      <td>0.011379</td>\n",
       "      <td>whole blood</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>533 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                title geo_accession                 status submission_date  \\\n",
       "GSM3667913    sample1    GSM3667913  Public on Mar 14 2019     Mar 13 2019   \n",
       "GSM3667872   sample10    GSM3667872  Public on Mar 14 2019     Mar 13 2019   \n",
       "GSM3667947  sample100    GSM3667947  Public on Mar 14 2019     Mar 13 2019   \n",
       "GSM3667878  sample101    GSM3667878  Public on Mar 14 2019     Mar 13 2019   \n",
       "GSM3667998  sample102    GSM3667998  Public on Mar 14 2019     Mar 13 2019   \n",
       "...               ...           ...                    ...             ...   \n",
       "GSM3667830   sample95    GSM3667830  Public on Mar 14 2019     Mar 13 2019   \n",
       "GSM3667866   sample96    GSM3667866  Public on Mar 14 2019     Mar 13 2019   \n",
       "GSM3667815   sample97    GSM3667815  Public on Mar 14 2019     Mar 13 2019   \n",
       "GSM3668001   sample98    GSM3668001  Public on Mar 14 2019     Mar 13 2019   \n",
       "GSM3667717   sample99    GSM3667717  Public on Mar 14 2019     Mar 13 2019   \n",
       "\n",
       "           last_update_date     type  channel_count source_name_ch1  \\\n",
       "GSM3667913      Mar 15 2019  genomic              1     whole blood   \n",
       "GSM3667872      Mar 15 2019  genomic              1     whole blood   \n",
       "GSM3667947      Mar 15 2019  genomic              1     whole blood   \n",
       "GSM3667878      Mar 15 2019  genomic              1     whole blood   \n",
       "GSM3667998      Mar 15 2019  genomic              1     whole blood   \n",
       "...                     ...      ...            ...             ...   \n",
       "GSM3667830      Mar 15 2019  genomic              1     whole blood   \n",
       "GSM3667866      Mar 15 2019  genomic              1     whole blood   \n",
       "GSM3667815      Mar 15 2019  genomic              1     whole blood   \n",
       "GSM3668001      Mar 15 2019  genomic              1     whole blood   \n",
       "GSM3667717      Mar 15 2019  genomic              1     whole blood   \n",
       "\n",
       "            organism_ch1 characteristics_ch1  ... cellcount.gran.ch1  \\\n",
       "GSM3667913  Homo sapiens     diagnosis: case  ...            0.57450   \n",
       "GSM3667872  Homo sapiens     diagnosis: case  ...            0.50000   \n",
       "GSM3667947  Homo sapiens     diagnosis: case  ...            0.52666   \n",
       "GSM3667878  Homo sapiens     diagnosis: case  ...            0.61128   \n",
       "GSM3667998  Homo sapiens     diagnosis: case  ...            0.53323   \n",
       "...                  ...                 ...  ...                ...   \n",
       "GSM3667830  Homo sapiens     diagnosis: case  ...            0.67893   \n",
       "GSM3667866  Homo sapiens     diagnosis: case  ...            0.49627   \n",
       "GSM3667815  Homo sapiens     diagnosis: case  ...            0.76167   \n",
       "GSM3668001  Homo sapiens     diagnosis: case  ...            0.72225   \n",
       "GSM3667717  Homo sapiens     diagnosis: case  ...            0.74141   \n",
       "\n",
       "           cellcount.mono.ch1 cellcount.nk.ch1 diagnosis.ch1  \\\n",
       "GSM3667913            0.11560          0.01708             1   \n",
       "GSM3667872            0.06131          0.00000             1   \n",
       "GSM3667947            0.10827          0.00085             1   \n",
       "GSM3667878            0.16026          0.02816             1   \n",
       "GSM3667998            0.05734          0.00000             1   \n",
       "...                       ...              ...           ...   \n",
       "GSM3667830            0.06394          0.02600             1   \n",
       "GSM3667866            0.06016          0.05033             1   \n",
       "GSM3667815            0.04776          0.02331             1   \n",
       "GSM3668001            0.08636          0.02902             1   \n",
       "GSM3667717            0.06519          0.04442             1   \n",
       "\n",
       "           education.level.ch1 history.of.myocardial.infarction.ch1 Sex.ch1  \\\n",
       "GSM3667913                 5.0                                  NaN       F   \n",
       "GSM3667872                 3.0                                  NaN       F   \n",
       "GSM3667947                 6.0                                  NaN       F   \n",
       "GSM3667878                 5.0                                  NaN       M   \n",
       "GSM3667998                 6.0                                  NaN       F   \n",
       "...                        ...                                  ...     ...   \n",
       "GSM3667830                 6.0                                  NaN       M   \n",
       "GSM3667866                 6.0                                  NaN       M   \n",
       "GSM3667815                 6.0                                  NaN       F   \n",
       "GSM3668001                 5.0                                  NaN       F   \n",
       "GSM3667717                 6.0                                  NaN       M   \n",
       "\n",
       "           snp.based.principle.component.1.ch1  \\\n",
       "GSM3667913                           -0.022566   \n",
       "GSM3667872                           -0.001741   \n",
       "GSM3667947                            0.007919   \n",
       "GSM3667878                            0.014514   \n",
       "GSM3667998                           -0.000620   \n",
       "...                                        ...   \n",
       "GSM3667830                            0.011203   \n",
       "GSM3667866                            0.014452   \n",
       "GSM3667815                           -0.003758   \n",
       "GSM3668001                           -0.000920   \n",
       "GSM3667717                           -0.009914   \n",
       "\n",
       "           snp.based.principle.component.2.ch1   tissue.ch1  \n",
       "GSM3667913                            0.016230  whole blood  \n",
       "GSM3667872                           -0.018000  whole blood  \n",
       "GSM3667947                            0.003596  whole blood  \n",
       "GSM3667878                            0.005688  whole blood  \n",
       "GSM3667998                           -0.007328  whole blood  \n",
       "...                                        ...          ...  \n",
       "GSM3667830                            0.002856  whole blood  \n",
       "GSM3667866                           -0.001970  whole blood  \n",
       "GSM3667815                            0.019465  whole blood  \n",
       "GSM3668001                           -0.010731  whole blood  \n",
       "GSM3667717                            0.011379  whole blood  \n",
       "\n",
       "[533 rows x 58 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pheno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5a3cef96",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "x = data.T\n",
    "#y = pheno['Title'].values.ravel() 42861\n",
    "y = pheno['diagnosis.ch1'].values.ravel()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "239a7c6f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((373, 326417), (160, 326417))"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "random.seed(42)\n",
    "np.random.seed(42)\n",
    "X_train, X_test, y_train, y_test= train_test_split(x,y,stratify=y,test_size=0.3, random_state=42)\n",
    "X_train.shape,X_test.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "82717c64",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now start training gene...\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "num = len(probeMapDict)\n",
    "gene_list = []\n",
    "print('Now start training gene...')\n",
    "\n",
    "data_train = X_train.T\n",
    "data_test = X_test.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a4c9c72",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "len(probeMapDict.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ad742c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!pip install --upgrade setuptools\n",
    "!pip install minepy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5056b975",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Test\n",
    "from __future__ import division\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "from sklearn import preprocessing\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.feature_selection import RFE\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.feature_selection import chi2\n",
    "#from elm import ELMClassifier\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "from sklearn.metrics import accuracy_score,f1_score,matthews_corrcoef,cohen_kappa_score,jaccard_score\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "import lightgbm as lgb\n",
    "import warnings\n",
    "from pycm import ConfusionMatrix\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import sys\n",
    "stdsc = StandardScaler()\n",
    "from sklearn.preprocessing import MinMaxScaler,StandardScaler\n",
    "scaler = MinMaxScaler()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7f046208",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import random\n",
    "random.seed(42)\n",
    "np.random.seed(42)\n",
    "def classfiers(X, Y,times,fold):\n",
    "    #Using six classifiers to do cross validation for X and Y times fold.\n",
    "    classifier = [SVC(kernel='linear'), GaussianNB(), KNeighborsClassifier(), DecisionTreeClassifier(random_state=10),\n",
    "                  GradientBoostingClassifier(random_state=10), \n",
    "                  LogisticRegression(),\n",
    "                  RandomForestClassifier(),\n",
    "                  XGBClassifier(),\n",
    "                  lgb.LGBMClassifier()\n",
    "                  ]\n",
    "    acc_res,f1_res,mcc_res,kappa_res,jaccard_res,csi_res = ([] for i in range(6))\n",
    "    for clf in classifier:\n",
    "        #if clf == classifier[5]:\n",
    "            #X = stdsc.fit_transform(X)\n",
    "        print(clf)\n",
    "        each_acc,each_f1,each_mcc,each_kappa,each_jaccard,each_csi = ([] for i in range(6))\n",
    "        for i in range(times):\n",
    "            acc_temp,f1_temp,mcc_temp,kappa_temp,jaccard_temp,csi_temp = ([] for i in range(6))\n",
    "            skf = StratifiedKFold(n_splits=fold, random_state=i, shuffle=True)\n",
    "            for train_index, test_index in skf.split(X, Y):\n",
    "                # print('Train: ',train_index,'Test: ',test_index)\n",
    "                X_train, X_test = X[train_index], X[test_index]\n",
    "                Y_train, Y_test = Y[train_index], Y[test_index]\n",
    "                clf.fit(X_train, Y_train)\n",
    "                #Different metrics\n",
    "                \n",
    "                acc = accuracy_score(Y_test, clf.predict(X_test))\n",
    "                #print(\"Accuracy...\"+str(acc))\n",
    "                \n",
    "                f1 = f1_score(Y_test, clf.predict(X_test))\n",
    "                #print(\"F1...\"+str(f1))\n",
    "                \n",
    "                mcc=matthews_corrcoef(Y_test, clf.predict(X_test))\n",
    "                #print(\"MCC...\"+str(mcc))\n",
    "                \n",
    "                \n",
    "                kappa=cohen_kappa_score(Y_test, clf.predict(X_test))\n",
    "                #print(\"kappa...\"+str(kappa))\n",
    "                \n",
    "                jaccard=jaccard_score(Y_test, clf.predict(X_test))\n",
    "                #print(\"jaccard...\"+str(jaccard))\n",
    "                #CSI\n",
    "               \n",
    "                cm = ConfusionMatrix(actual_vector=Y_test, predict_vector=clf.predict(X_test))\n",
    "\n",
    "                #cm.print_matrix()\n",
    "                csi=cm.CSI\n",
    "                #print(csi)\n",
    "                #print(type(csi))\n",
    "                csiF= 0 if csi is \"None\" else csi\n",
    "                #print(csiF)\n",
    "                ###################\n",
    "                acc_temp.append(acc)\n",
    "                f1_temp.append(f1)\n",
    "                mcc_temp.append(mcc)\n",
    "                kappa_temp.append(kappa)\n",
    "                jaccard_temp.append(jaccard)\n",
    "                csi_temp.append(csiF)\n",
    "            each_acc.append(np.mean(acc_temp))\n",
    "            each_f1.append(np.mean(f1_temp))\n",
    "            each_mcc.append(np.mean(mcc_temp))\n",
    "            each_kappa.append(np.mean(kappa_temp))\n",
    "            each_jaccard.append(np.mean(jaccard_temp))\n",
    "            each_csi.append(np.mean(csi_temp))\n",
    "            \n",
    "        acc_res.append(np.mean(each_acc))\n",
    "        f1_res.append(np.mean(each_f1))\n",
    "        mcc_res.append(np.mean(each_mcc))\n",
    "        kappa_res.append(np.mean(each_kappa))\n",
    "        jaccard_res.append(np.mean(each_jaccard)) \n",
    "        csi_res.append(np.mean(each_csi))\n",
    "    return {\"acc\":acc_res,\"f1\":f1_res,\"mcc\":mcc_res,\"kappa\":kappa_res,\"jaccard\":jaccard_res,\"csi\":csi_res}\n",
    "\n",
    "def IFS_validation(X, Y, times,fold,tag):\n",
    "    feature_order = X.columns\n",
    "    list_gene = []\n",
    "    result={}\n",
    "    result_acc,result_f1,result_mcc,result_kappa,result_jaccard,result_csi = ([] for i in range(6))\n",
    "    print(\"feature_number:\")\n",
    "\n",
    "    result=classfiers(X.values, Y,times,fold)\n",
    "    result_acc.append(result[\"acc\"])\n",
    "    result_f1.append(result[\"f1\"])\n",
    "    result_mcc.append(result[\"mcc\"])\n",
    "    result_kappa.append(result[\"kappa\"])\n",
    "    result_jaccard.append(result[\"jaccard\"])\n",
    "    result_csi.append(result[\"csi\"])\n",
    "        \n",
    "    print(\"saving acc\")    \n",
    "    score = pd.DataFrame(result_acc)\n",
    "    col = ['SVM', 'NBayes', 'KNN', 'DTree', 'GBDT','LogisticRegression', 'RF', 'xgboost','lightGBM']\n",
    "    score.columns = col\n",
    "    #score.index = list(map(lambda x: ','.join(x), list_gene))\n",
    "    score.to_csv(tag + \"_acc_\" + str(times) + \"_runs_\" + str(fold) + \"_fold.csv\")\n",
    "    \n",
    "    \n",
    "    print(\"saving f1\")\n",
    "    score = pd.DataFrame(result_f1)\n",
    "    score.columns = col\n",
    "    #score.index = list(map(lambda x: ','.join(x), list_gene))\n",
    "    score.to_csv(tag  + \"_f1_\" + str(times) + \"_runs_\" + str(fold) + \"_fold.csv\")\n",
    "    \n",
    "    print(\"saving mcc\")\n",
    "    score = pd.DataFrame(result_mcc)\n",
    "    score.columns = col\n",
    "    #score.index = list(map(lambda x: ','.join(x), list_gene))\n",
    "    score.to_csv(tag  + \"_mcc_\" + str(times) + \"_runs_\" + str(fold) + \"_fold.csv\")\n",
    "    \n",
    "    print(\"saving kappa\")\n",
    "    score = pd.DataFrame(result_kappa)\n",
    "    score.columns = col\n",
    "    #score.index = list(map(lambda x: ','.join(x), list_gene))\n",
    "    score.to_csv(tag  + \"_kappa_\" + str(times) + \"_runs_\" + str(fold) + \"_fold.csv\")\n",
    "    \n",
    "    print(\"saving jaccard\")\n",
    "    score = pd.DataFrame(result_jaccard)\n",
    "    score.columns = col\n",
    "    #score.index = list(map(lambda x: ','.join(x), list_gene))\n",
    "    score.to_csv(tag + \"_jaccard_\" + str(times) + \"_runs_\" + str(fold) + \"_fold.csv\")\n",
    "    \n",
    "    print(\"saving csi\")\n",
    "    score = pd.DataFrame(result_csi)\n",
    "    score.columns = col\n",
    "    #score.index = list(map(lambda x: ','.join(x), list_gene))\n",
    "    score.to_csv(tag  + \"_csi_\" + str(times) + \"_runs_\" + str(fold) + \"_fold.csv\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "46977034",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import random\n",
    "random.seed(42)\n",
    "np.random.seed(42)\n",
    "def select_feature(code,data,label,gene,times=20,fold=5,cnt=100):\n",
    "    if gene:\n",
    "        geneOrResidue=\"gene\"\n",
    "    else:\n",
    "        geneOrResidue=\"site\"\n",
    "\n",
    "    tag = code + \"_\" + geneOrResidue + \"_\" + model_type # Prefix for result file\n",
    "    #methods\n",
    "    IFS_validation(data, label,times,fold,tag)\n",
    "    \n",
    "def load_data(filename, labelname):\n",
    "    data = pd.read_table(filename, index_col=0)\n",
    "    label = labelname\n",
    "    return data, label\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "208d32ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "160\n",
      "19908\n",
      "(160, 19908)\n",
      "feature_number:\n",
      "SVC(kernel='linear')\n",
      "GaussianNB()\n",
      "KNeighborsClassifier()\n",
      "DecisionTreeClassifier(random_state=10)\n",
      "GradientBoostingClassifier(random_state=10)\n",
      "LogisticRegression()\n",
      "RandomForestClassifier()\n",
      "XGBClassifier(base_score=None, booster=None, callbacks=None,\n",
      "              colsample_bylevel=None, colsample_bynode=None,\n",
      "              colsample_bytree=None, early_stopping_rounds=None,\n",
      "              enable_categorical=False, eval_metric=None, gamma=None,\n",
      "              gpu_id=None, grow_policy=None, importance_type=None,\n",
      "              interaction_constraints=None, learning_rate=None, max_bin=None,\n",
      "              max_cat_to_onehot=None, max_delta_step=None, max_depth=None,\n",
      "              max_leaves=None, min_child_weight=None, missing=nan,\n",
      "              monotone_constraints=None, n_estimators=100, n_jobs=None,\n",
      "              num_parallel_tree=None, predictor=None, random_state=None,\n",
      "              reg_alpha=None, reg_lambda=None, ...)\n",
      "LGBMClassifier()\n",
      "saving acc\n",
      "saving f1\n",
      "saving mcc\n",
      "saving kappa\n",
      "saving jaccard\n",
      "saving csi\n",
      "160\n",
      "326417\n",
      "feature_number:\n",
      "SVC(kernel='linear')\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_11640/57380478.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     29\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     30\u001b[0m \u001b[0mlabel\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0my_test\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 31\u001b[1;33m \u001b[0mselect_feature\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcode\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mlabel\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mgene\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mFalse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_11640/4090642886.py\u001b[0m in \u001b[0;36mselect_feature\u001b[1;34m(code, data, label, gene, times, fold, cnt)\u001b[0m\n\u001b[0;32m     10\u001b[0m     \u001b[0mtag\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcode\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m\"_\"\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mgeneOrResidue\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m\"_\"\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mmodel_type\u001b[0m \u001b[1;31m# Prefix for result file\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m     \u001b[1;31m#methods\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 12\u001b[1;33m     \u001b[0mIFS_validation\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtimes\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mfold\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtag\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     13\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     14\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mload_data\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlabelname\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_11640/3519347790.py\u001b[0m in \u001b[0;36mIFS_validation\u001b[1;34m(X, Y, times, fold, tag)\u001b[0m\n\u001b[0;32m     81\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"feature_number:\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     82\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 83\u001b[1;33m     \u001b[0mresult\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mclassfiers\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mY\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtimes\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mfold\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     84\u001b[0m     \u001b[0mresult_acc\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"acc\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     85\u001b[0m     \u001b[0mresult_f1\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"f1\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_11640/3519347790.py\u001b[0m in \u001b[0;36mclassfiers\u001b[1;34m(X, Y, times, fold)\u001b[0m\n\u001b[0;32m     37\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     38\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 39\u001b[1;33m                 \u001b[0mkappa\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcohen_kappa_score\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mY_test\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mclf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX_test\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     40\u001b[0m                 \u001b[1;31m#print(\"kappa...\"+str(kappa))\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     41\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\sklearn\\svm\\_base.py\u001b[0m in \u001b[0;36mpredict\u001b[1;34m(self, X)\u001b[0m\n\u001b[0;32m    808\u001b[0m             \u001b[0my\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0margmax\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdecision_function\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    809\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 810\u001b[1;33m             \u001b[0my\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0msuper\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    811\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclasses_\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mintp\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    812\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\sklearn\\svm\\_base.py\u001b[0m in \u001b[0;36mpredict\u001b[1;34m(self, X)\u001b[0m\n\u001b[0;32m    433\u001b[0m         \u001b[0mX\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_validate_for_predict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    434\u001b[0m         \u001b[0mpredict\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_sparse_predict\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_sparse\u001b[0m \u001b[1;32melse\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_dense_predict\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 435\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mpredict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    436\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    437\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_dense_predict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mX\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\sklearn\\svm\\_base.py\u001b[0m in \u001b[0;36m_dense_predict\u001b[1;34m(self, X)\u001b[0m\n\u001b[0;32m    452\u001b[0m         \u001b[0msvm_type\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mLIBSVM_IMPL\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_impl\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    453\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 454\u001b[1;33m         return libsvm.predict(\n\u001b[0m\u001b[0;32m    455\u001b[0m             \u001b[0mX\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    456\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msupport_\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#check performance for whole data\n",
    "###################################################################\n",
    "code=\"GSE128235\"\n",
    "model_type=\"sUMAP\"\n",
    "filename=code+\"_gene_level_test(\"+model_type+\").txt\"\n",
    "labelname=y_test\n",
    "\n",
    "data, label = load_data(filename, labelname)\n",
    "\n",
    "dataT=data.T\n",
    "np.random.seed(1)\n",
    "\n",
    "print(dataT.shape[0])\n",
    "print(dataT.shape[1])\n",
    "\n",
    "scalar=MinMaxScaler()\n",
    "dataT=dataT.dropna() \n",
    "data1=pd.DataFrame(scalar.fit_transform(dataT),columns=data.index,index=data.columns)\n",
    "print(data1.shape)\n",
    "\n",
    "model_type=\"sUMAP-Whole\"\n",
    "select_feature(code,data1,label,gene=True)\n",
    "\n",
    "#site level\n",
    "#####################################################################\n",
    "model_type=\"site-Whole\"\n",
    "data=data_test.T\n",
    "print(data.shape[0])\n",
    "print(data.shape[1])\n",
    "label=y_test\n",
    "select_feature(code,data,label,gene=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7869281d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19908\n",
      "160\n",
      "feature_number:\n",
      "SVC(kernel='linear')\n",
      "GaussianNB()\n",
      "KNeighborsClassifier()\n",
      "DecisionTreeClassifier(random_state=10)\n",
      "GradientBoostingClassifier(random_state=10)\n",
      "LogisticRegression()\n",
      "RandomForestClassifier()\n",
      "XGBClassifier(base_score=None, booster=None, callbacks=None,\n",
      "              colsample_bylevel=None, colsample_bynode=None,\n",
      "              colsample_bytree=None, early_stopping_rounds=None,\n",
      "              enable_categorical=False, eval_metric=None, gamma=None,\n",
      "              gpu_id=None, grow_policy=None, importance_type=None,\n",
      "              interaction_constraints=None, learning_rate=None, max_bin=None,\n",
      "              max_cat_to_onehot=None, max_delta_step=None, max_depth=None,\n",
      "              max_leaves=None, min_child_weight=None, missing=nan,\n",
      "              monotone_constraints=None, n_estimators=100, n_jobs=None,\n",
      "              num_parallel_tree=None, predictor=None, random_state=None,\n",
      "              reg_alpha=None, reg_lambda=None, ...)\n",
      "LGBMClassifier()\n",
      "saving acc\n",
      "saving f1\n",
      "saving mcc\n",
      "saving kappa\n",
      "saving jaccard\n",
      "saving csi\n",
      "160\n",
      "1624\n",
      "feature_number:\n",
      "SVC(kernel='linear')\n",
      "GaussianNB()\n",
      "KNeighborsClassifier()\n",
      "DecisionTreeClassifier(random_state=10)\n",
      "GradientBoostingClassifier(random_state=10)\n",
      "LogisticRegression()\n",
      "RandomForestClassifier()\n",
      "XGBClassifier(base_score=None, booster=None, callbacks=None,\n",
      "              colsample_bylevel=None, colsample_bynode=None,\n",
      "              colsample_bytree=None, early_stopping_rounds=None,\n",
      "              enable_categorical=False, eval_metric=None, gamma=None,\n",
      "              gpu_id=None, grow_policy=None, importance_type=None,\n",
      "              interaction_constraints=None, learning_rate=None, max_bin=None,\n",
      "              max_cat_to_onehot=None, max_delta_step=None, max_depth=None,\n",
      "              max_leaves=None, min_child_weight=None, missing=nan,\n",
      "              monotone_constraints=None, n_estimators=100, n_jobs=None,\n",
      "              num_parallel_tree=None, predictor=None, random_state=None,\n",
      "              reg_alpha=None, reg_lambda=None, ...)\n",
      "LGBMClassifier()\n",
      "saving acc\n",
      "saving f1\n",
      "saving mcc\n",
      "saving kappa\n",
      "saving jaccard\n",
      "saving csi\n"
     ]
    }
   ],
   "source": [
    "#Randomly select 100 genes and try both sUAGM and site level\n",
    "code=\"GSE128235\"\n",
    "model_type=\"sUMAP\"\n",
    "filename=code+\"_gene_level_test(\"+model_type+\").txt\"\n",
    "labelname=y_test\n",
    "\n",
    "data, label = load_data(filename, labelname)\n",
    "\n",
    "dataT=data.T\n",
    "np.random.seed(1)\n",
    "print(dataT.shape[1])\n",
    "print(dataT.shape[0])\n",
    "\n",
    "N_columns =  100\n",
    "# Given the number of features, generate features without replacement\n",
    "columns = list(np.random.choice(range(dataT.shape[1]), N_columns, replace=False))\n",
    "    \n",
    "scalar=MinMaxScaler()\n",
    "dataT=dataT.dropna() \n",
    "data1=pd.DataFrame(scalar.fit_transform(dataT.iloc[:,columns]),columns=data.index[columns],index=dataT.index)\n",
    "genes=data.index[columns]\n",
    "model_type=\"sUMAP-Random100\"\n",
    "select_feature(code,data1,label,gene=True)\n",
    "\n",
    "#site level\n",
    "#####################################################################\n",
    "model_type=\"site-Random100\"\n",
    "count=0\n",
    "gene_data = []\n",
    "for gene in genes:\n",
    "    count += 1\n",
    "    for residue in data_test.index:    \n",
    "        if residue in probeMapDict[gene]:\n",
    "            gene_data.append(data_test.loc[residue])\n",
    "    if len(gene_data) == 0:\n",
    "        print('Contained Nan data, has been removed!')\n",
    "        continue\n",
    "gene_data = pd.DataFrame(np.array(gene_data))\n",
    "data=gene_data.T\n",
    "print(data.shape[0])\n",
    "print(data.shape[1])\n",
    "label=y_test\n",
    "select_feature(code,data,label,gene=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54790870",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
